compiler/modules/front/const.bal

```diff
@@ -517,47 +520,88 @@ function foldVarRefExpr(FoldContext cx, t:SemType? expectedType, s:VarRefExpr ex
+        if expectedType == () || expectsFloat(tx, expectedType) {
+            result = floatFromFpLiteral(untypedLiteral);
+        }
+        else if expectsDecimal(tx, expectedType) {
+            result = check decimalFromFpLiteral(cx, untypedLiteral, startPos);
+        }
+        else {
+            result = floatFromFpLiteral(untypedLiteral);
+        }
     }
+    return { startPos: startPos, endPos: expr.endPos, value: result };
 }
 
 function foldIntLiteralExpr(FoldContext cx, t:SemType? expectedType, s:IntLiteralExpr expr) returns s:ConstValueExpr|FoldError {
-    float|int|error result;
-    string ty;
-    if expr.base == 10 && expectsFloat(expectedType) {
-        result = floatFromDecimalLiteral(expr.digits);
-        ty = "float"; 
+    int|float|decimal result;
+    t:Context tx = cx.typeContext();
+    Position startPos = expr.startPos;
+    if expectedType == () || expectsInt(tx, expectedType) {
+        result = check intFromLiteral(cx, expr);
+    }  
+    else if expectsFloat(tx, expectedType) {
+        result = floatFromFpLiteral(expr.digits);
+    }
+    else if expectsDecimal(tx, expectedType) {
+        result = check decimalFromFpLiteral(cx, expr.digits, startPos);
     }
     else {
-        result = s:intFromIntLiteral(expr.base, expr.digits);
-        ty = "int";
-    }
-    if result is int|float {
-        return { startPos: expr.startPos, endPos: expr.endPos, value: result };
-    }
-    else {
-        return cx.semanticErr("invalid " + ty + " literal", cause=result, pos=expr.startPos);
+        result = check intFromLiteral(cx, expr);
     }
+    return { startPos: startPos, endPos: expr.endPos, value: result };
+}
+
+function expectsInt(t:Context cx, t:SemType semType) returns boolean {
+    return !t:isEmpty(cx, t:intersect(semType, t:INT));
+}
+
+function expectsFloat(t:Context cx, t:SemType semType) returns boolean {
+    return !t:isEmpty(cx, t:intersect(semType, t:FLOAT));
+}
+
+function expectsDecimal(t:Context cx, t:SemType semType) returns boolean {
+    return !t:isEmpty(cx, t:intersect(semType, t:DECIMAL));
 }
 
-function expectsFloat(t:SemType? semType) returns boolean {
-    if semType == () {
-        return false;
+// Since the binary floating point literal is parsed correctly,
+// it is impossible to throw an error.
+function floatFromFpLiteral(string digits) returns float {
+    float|error f = float:fromString(digits);
+    if f is error {
+        panic err:impossible();
     }
-    else {
-        return t:isSubtypeSimple(t:intersect(semType, t:union(t:FLOAT, t:INT)), t:FLOAT);
+    return f;
+}
+
+// Even if the decimal floating point literal is parsed correctly,
+// infinities should through an error.
+function decimalFromFpLiteral(FoldContext cx, string decimalStr, Position pos) returns decimal|FoldError {
+    decimal|error d = decimal:fromString(decimalStr);
+    if d is error {
+        return cx.semanticErr("invalid decimal literal", pos);
```

Comment:
For what input will this happen?

Reply:
For values greater than decimal max or less than min(ex: 1e6145, -1e6145), these values are parsed correctly, but fromString should give an error(https://github.com/ballerina-platform/ballerina-lang/issues/33805#issuecomment-988473730).

Reply:
Does this mean large float values don't give an error and large decimals do? that is odd.

Reply:
Ahh this is because we don't have decimals infinity.